File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        heap = []
        ans = 0
        n = len(matrix)
        m = len(matrix[0])
               
        for i in range(len(matrix[0])):
            heapq.heappush(heap,(matrix[0][i],0,i))
                           
        for i in range(k,0,-1):
            ans = heapq.heappop(heap)
            p = ans[1] + 1
            q = ans[2]
            if p < m:
                val = matrix[p][ans[2]] 
                heapq.heappush(heap,(val,p,q))
                
        return ans[0]
            
        
        

File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length; 
        int left = matrix[0][0], right = matrix[n-1][n-1];
        
        while (left < right){
            int count = 0;
            int mid = left + (right - left ) / 2;
            int j = n - 1;
            for (int i = 0; i < n; i++){
                while (j >= 0 && matrix[i][j] > mid) j--;
                count +=  (j + 1);
            }
            
            if (count >= k) right = mid;
            else left = mid + 1;
        }
        return left;
    }
}